apollo-spaceport/src/lib.rs

```diff
@@ -0,0 +1,226 @@
+    fn from(error: std::io::Error) -> Self {
+        ReporterError {
+            msg: error.to_string(),
+            source: Box::new(error),
+        }
+    }
+}
+
+impl From<sys_info::Error> for ReporterError {
+    fn from(error: sys_info::Error) -> Self {
+        ReporterError {
+            msg: error.to_string(),
+            source: Box::new(error),
+        }
+    }
+}
+
+impl From<JoinError> for ReporterError {
+    fn from(error: JoinError) -> Self {
+        ReporterError {
+            msg: error.to_string(),
+            source: Box::new(error),
+        }
+    }
+}
+
+impl std::fmt::Display for ReporterError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(
+            f,
+            "ReporterError: source: {}, message: {}",
+            self.source, self.msg
+        )
+    }
+}
+
+impl Eq for ReporterGraph {}
+
+// PartialEq is derived in the generated code, but Hash isn't and we need
+// it to use this as key in a HashMap. We have to make sure this
+// implementation always matches the derived PartialEq in the generated
+// code.
+#[allow(clippy::derive_hash_xor_eq)]
+impl Hash for ReporterGraph {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.reference.hash(state);
+        self.key.hash(state);
+    }
+}
+
+impl Report {
+    /// Try to create a new Report.
+    ///
+    /// This can fail if the ReportHeader is not created.
+    pub fn try_new(graph: &str) -> Result<Self, ReporterError> {
+        let header = Some(ReportHeader::try_new(graph)?);
+
+        Ok(Report {
+            header,
+            ..Default::default()
+        })
+    }
+}
+
+#[cfg(target_os = "windows")]
+fn get_uname() -> Result<String, std::io::Error> {
+    // Best we can do on windows right now
+    let sysname = sys_info::os_type().unwrap_or("Windows".to_string());
+    let nodename = sys_info::hostname().unwrap_or("unknown".to_string());
+    let release = sys_info::os_release().unwrap_or("unknown".to_string());
+    let version = "unknown";
+    let machine = "unknown";
+    Ok(format!(
+        "{}, {}, {}, {}, {}",
+        sysname, nodename, release, version, machine
+    ))
+}
+
+#[cfg(not(target_os = "windows"))]
+fn get_uname() -> Result<String, std::io::Error> {
+    let u = uname::uname()?;
+    Ok(format!(
+        "{}, {}, {}, {}, {},",
+        u.sysname, u.nodename, u.release, u.version, u.machine
+    ))
+}
+
+impl ReportHeader {
+    fn try_new(graph: &str) -> Result<Self, ReporterError> {
+        let mut header = ReportHeader {
+            ..Default::default()
+        };
+
+        header.hostname = hostname()?;
+        header.agent_version = format!(
+            "{}-{}",
+            std::env!("CARGO_PKG_NAME"),
+            std::env!("CARGO_PKG_VERSION")
+        );
+        header.runtime_version = "rust".to_string();
+        header.uname = get_uname()?;
+        header.graph_ref = graph.to_string();
+        Ok(header)
```

Comment:
This is notably missing executable_schema_id (a hash of a schema somehow), which is fine for now but worth a comment.

Reply:
I'll add an issue. Is this something the enhanced router_bridge interface could report?

Reply:
Yeah, probably. It will probably be challenging to implement accurately in router in the same way that gateway/server does, because that's "use `graphql-js`'s schema printing function to print the schema (with whatever ordering/formatting/etc choices it happens to make) and then takes a hash of it". It's possible that uplink could (or does?) send down a similar/identical hash though?

I have to say that I do not really understand to what degree the ReportHeader.executable_schema_id usage reporting field is actually used in our system. I know it does end up in some Druid columns but I'm not sure what if anything it's used for end-to-end. I also am not sure if this hash ends up needing to match with the identifier for schemas in the schema registry in any way (I never really learned about how the schema registry works). I think @zionts and/or @jsegaran have explained this to me in the past but I don't remember.